### PR TITLE
CLOUDP-212054: Sign images with Garasign

### PR DESCRIPTION
**Pull request by josvazg**
✅ [Sign and self verification test passed](https://github.com/mongodb/mongodb-atlas-kubernetes/actions/runs/8003743939/job/21859696852)

Includes:
- Signing images for releases and rebuilds.
- Signing from Makefile rule.
- Script to detect multi architecture SHAs and sing the parent manifest and platform images.
- Script for each individual image signature.

Note the new secrets required and already populated in the CI `release` environment:
- Parameter `PKCS11_URI` required by Garasign
- Credentials for Garasign: `GRS_USERNAME` & `GRS_PASSWORD`
- Credentials for the MongoDB Artifactory holding the Garasign container `MDB_ARTIFACTORY_USERNAME` & `MDB_ARTIFACTORY_PASSWORD`

### All Submissions:

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?

### CODE REVIEW

File: .github/workflows/rebuild-released-images.yaml [hunk 1]
```diff
@@ -52,6 +52,7 @@ jobs:
           fi
   
   build-and-publish-image:
+    environment: release
```
Comment (slaskawi) [inline .github/workflows/rebuild-released-images.yaml:55]: Do we support signing in daily rebuilds?

Comment (josvazg) [reply .github/workflows/rebuild-released-images.yaml:55]: Yes, it is the other yaml, `rebuild-released-images.yaml. But that one I cannot test until I merge it. Code is mostly the same as for a regular release.

Comment (slaskawi) [reply .github/workflows/rebuild-released-images.yaml:55]: Perfect! thanks!

**Review by slaskawi — APPROVED** [review]
LGTM but would be great to have another AKO engineer to look at this.

File: scripts/sign-multiarch.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+docker pull "${img_to_sign}"
+MULTIARCH_IMG_SHA=$(docker inspect "${img_to_sign}" |jq -rc '.[0].Id')
+IMG_PLATFORMS_SHAS=$(docker manifest inspect "${img_to_sign}" | \
+  jq -rc '.manifests[] | select(.platform.os != "unknown" and .platform.architecture != "unknown") | .digest')
+
+echo "Signing parent multiarch image ${img_to_sign}@${MULTIARCH_IMG_SHA}..."
+IMG_TO_SIGN="${img_to_sign}@${MULTIARCH_IMG_SHA}" "${SCRIPT_DIR}"/sign.sh
+
+for platform_sha in ${IMG_PLATFORMS_SHAS}; do
+  echo "Signing platform image ${img_to_sign}@${platform_sha}..."
```
Comment (s-urbaniak) [inline scripts/sign-multiarch.sh:19]: nit: whitespaces vs. tabs in this file, let's be consistent

File: scripts/sign.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
+set -euo pipefail
+
+REPO=${IMG_REPO:-docker.io/mongodb/mongodb-atlas-kubernetes-operator-prerelease}
+img_to_sign=${IMG_TO_SIGN:-$REPO:$VERSION}
+SIGNATURE_REPO=${SIGNATURE_REPO:-$REPO}
+#SIGNATURE_REPO=docker.io/mongodb/signatures
+#SIGNATURE_REPO=quay.io/mongodb/signatures
```
Comment (s-urbaniak) [inline scripts/sign.sh:9]: remove commented code

File: scripts/sign.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+DOCKER_CFG=${DOCKER_CFG:-~/.docker/config.json}
+
+SIGNING_ENVFILE="${TMPDIR}/signing-envfile"
+
+{
+  echo "GRS_CONFIG_USER1_USERNAME=${GRS_USERNAME}";
+  echo "GRS_CONFIG_USER1_PASSWORD=${GRS_PASSWORD}";
+  echo "COSIGN_REPOSITORY=${SIGNATURE_REPO}";
+}  >> "${SIGNING_ENVFILE}"
```
Comment (s-urbaniak) [inline scripts/sign.sh:27]: why not `>` ?

**Review by s-urbaniak — COMMENTED** [review]
just nits, else lgtm

File: scripts/sign-multiarch.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+set -euo pipefail
+
+REPO=${IMG_REPO:-docker.io/mongodb/mongodb-atlas-kubernetes-operator-prerelease}
```
Comment (roothorp) [inline scripts/sign-multiarch.sh:5]: Where is `IMG_REPO` coming from?

File: scripts/verify.sh [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+set -euo pipefail
+
+REPO=${IMG_REPO:-docker.io/mongodb/mongodb-atlas-kubernetes-operator-prerelease}
+img_to_verify=${IMG:-$REPO:$VERSION}
+SIGNATURE_REPO=${SIGNATURE_REPO:-$REPO}
+
+KEY_FILE=${KEY_FILE:-ako.pem}
+
+COSIGN_REPOSITORY="${SIGNATURE_REPO}" cosign verify \
+  --key="${KEY_FILE}" "${img_to_verify}" # --insecure-ignore-tlog
```
[the remark this answers is not in the document]
Comment (josvazg) [reply scripts/verify.sh:12]: yep, will remove it. Thanks!

File: scripts/sign-multiarch.sh (hunk 5 above)
Comment (josvazg) [reply scripts/sign-multiarch.sh:5]: Is a env var input. Having `set -euo` above will make the script fail if we omit the env var.

File: scripts/verify.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (josvazg) [reply scripts/verify.sh]: Done

**Review by helderjs — APPROVED** [review]

**Pull request merged**